### PR TITLE
Refactored resources usage in AssetManager

### PR DESCRIPTION
**Pull request by teofilobd**
Refactored `AssetManager` to hold resources using `shared_ptr` instead of raw pointers. Then, fixed all other classes affected by the changes above.

### CODE REVIEW

File: annileen/engine/assetmanager.ixx [hunk 1]
```diff
@@ -28,14 +28,14 @@ export namespace annileen
 	class AssetManager final
 	{
 	private:
-		std::map<std::string, AssetTableEntry> m_Assets;
+		std::map<std::string, std::shared_ptr<AssetTableEntry>> m_Assets;
 		AssetWatcher* m_Watcher;
```
Comment (crocidb) [inline annileen/engine/assetmanager.ixx]: worth adding the watcher as well?

File: annileen/engine/assetmanager.ixx [hunk 2]
```diff
@@ -367,18 +356,18 @@ namespace annileen
...
 	}
 
-	Cubemap* AssetManager::getCubemap(const std::string& name)
+	std::shared_ptr<Cubemap> AssetManager::getCubemap(const std::string& name)
 	{
 		auto entry = getAssetEntry(name);
 		if (entry->m_Loaded)
 		{
-			return static_cast<Cubemap*>(entry->m_Asset);
+			return dynamic_pointer_cast<Cubemap>(entry->m_Asset);
```
Comment (crocidb) [inline annileen/engine/assetmanager.ixx]: is there any static cast option?

File: annileen/engine/asset.ixx [hunk 3]
```diff
@@ -25,11 +26,11 @@ export namespace annileen
...
 	struct AssetTableEntry
 	{
-		std::string m_Filepath;
-		AssetType m_Type;
+		std::string m_Filepath{""};
+		AssetType m_Type{ AssetType::Undefined };
 
-		bool m_Loaded;
-		AssetObject* m_Asset;
+		bool m_Loaded{ false };
```
Comment (crocidb) [inline annileen/engine/asset.ixx:32]: are those initialization happening on the default constructor?

File: annileen/engine/cubemap.ixx [hunk 4]
```diff
@@ -40,5 +40,7 @@ namespace annileen
 
 	Cubemap::~Cubemap()
 	{
+		// TODO: remove
+		std::cout << "Cubemap destroyed." << std::endl;
```
Comment (crocidb) [inline annileen/engine/cubemap.ixx:44]: is that supposed to be removed?

File: annileen/engine/mesh.ixx [hunk 5]
```diff
@@ -76,20 +76,19 @@ namespace annileen
...
-	Mesh::Mesh()
-	{
-	}
-
 	Mesh::~Mesh()
 	{
 		unload();
+
+		// TODO: remove
+		std::cout << "Mesh destroyed." << std::endl;
```
Comment (crocidb) [inline annileen/engine/mesh.ixx:84]: same here.

File: annileen/engine/mesh.ixx [hunk 6]
```diff
@@ -76,20 +76,19 @@ namespace annileen
...
 	MeshGroup::~MeshGroup()
 	{
-		for (auto& m : m_Meshes)
-		{
-			delete m;
-		}
+		m_Meshes.clear();
+
+		// TODO: remove
+		std::cout << "MeshGroup destroyed." << std::endl;
```
Comment (crocidb) [inline annileen/engine/mesh.ixx:92]: hehe here too. also, why not `ANNI_LOG` instead of `std::cout`?

**Review by crocidb — CHANGES_REQUESTED** [review]

File: annileen/engine/assetmanager.ixx (hunk 1 above)
Comment (teofilobd) [reply annileen/engine/assetmanager.ixx]: yes. I'll do.

File: annileen/engine/assetmanager.ixx (hunk 2 above)
Comment (teofilobd) [reply annileen/engine/assetmanager.ixx]: There is static_pointer_cast, but why not dynamic?

File: annileen/engine/asset.ixx (hunk 3 above)
Comment (teofilobd) [reply annileen/engine/asset.ixx:32]: Yes.

File: annileen/engine/cubemap.ixx (hunk 4 above)
Comment (teofilobd) [reply annileen/engine/cubemap.ixx:44]: Yes. I'm adding these comments for debugging reasons for now. I'll go over those TODOs later and delete them.

File: annileen/engine/mesh.ixx (hunk 5 above)
Comment (teofilobd) [reply annileen/engine/mesh.ixx:84]: I'll delete later. Can be merged.

File: annileen/engine/mesh.ixx (hunk 6 above)
Comment (teofilobd) [reply annileen/engine/mesh.ixx:92]: I'll delete later. Can be merged. It was just for debug, but could use anni_log.

File: annileen/engine/assetmanager.ixx (hunk 2 above)
Comment (crocidb) [reply annileen/engine/assetmanager.ixx]: i don't know exactly, just imagine static has less overhead, and if it worked with it, maybe we should keep it?

**Review by crocidb — APPROVED** [review]

File: annileen/engine/assetmanager.ixx (hunk 2 above)
Comment (teofilobd) [reply annileen/engine/assetmanager.ixx]: Changed it.

**Pull request merged**